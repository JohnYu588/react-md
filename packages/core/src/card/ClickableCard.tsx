"use client";

import { cnb } from "cnbuilder";
import { type MouseEventHandler, forwardRef } from "react";

import { type ComponentWithRippleProps } from "../interaction/types.js";
import { useElementInteraction } from "../interaction/useElementInteraction.js";
import { useHigherContrastChildren } from "../interaction/useHigherContrastChildren.js";
import { Card, type CardProps } from "./Card.js";

const noop = (): void => {
  // do nothing
};

/**
 * @since 6.0.0
 */
export interface ClickableCardProps
  extends CardProps,
    ComponentWithRippleProps {
  onClick: MouseEventHandler<HTMLDivElement>;

  /**
   * Set this to `true` to disable all click events.
   *
   * Note: Any disabled styling will need to be manually applied.
   *
   * @defaultValue `false`
   */
  disabled?: boolean;
}

/**
 * A Small wrapper around the `Card` component that is clickable and has the
 * element interaction enabled (ripples).
 *
 * @example Simple Example
 * ```tsx
 * import { CardContent } from "@react-md/core/card/CardContent"
 * import { ClickableCard } from "@react-md/core/card/ClickableCard";
 * import { type ReactElement } from "react";
 *
 * export default function ClickableCardExample(): ReactElement {
 *   return (
 *     <ClickableCard
 *       onClick={() => {
 *         // do something
 *       }}
 *     >
 *       <CardContent>Wow</CardContent>
 *     </ClickableCard>
 *   );
 * }
 * ```
 *
 * @since 6.0.0
 */
export const ClickableCard = forwardRef<HTMLDivElement, ClickableCardProps>(
  function ClickableCard(props, ref) {
    const {
      role = "button",
      disabled,
      tabIndex = disabled ? undefined : 0,
      onClick,
      onKeyDown = noop,
      className,
      children: propChildren,
      disableRipple,
      ...remaining
    } = props;
    const { handlers, ripples } = useElementInteraction({
      ...remaining,
      mode: disableRipple ? "none" : undefined,
    });
    const children = useHigherContrastChildren(propChildren);

    return (
      <Card
        {...remaining}
        {...handlers}
        aria-disabled={disabled || undefined}
        ref={ref}
        role={role}
        tabIndex={tabIndex}
        className={cnb("rmd-card--clickable", className)}
        onClick={(event) => {
          if (disabled) {
            return;
          }

          onClick(event);
        }}
        onKeyDown={(event) => {
          onKeyDown(event);
          if (disabled) {
            return;
          }

          const isSpace = event.key === " ";
          if (isSpace || event.key === "Enter") {
            if (isSpace) {
              event.preventDefault();
            }

            event.currentTarget.click();
          }
        }}
      >
        {children}
        {ripples}
      </Card>
    );
  }
);
