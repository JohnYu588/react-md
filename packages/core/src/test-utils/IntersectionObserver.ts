export class IntersectionObserverMock implements IntersectionObserver {
  root: Document | Element | null;
  rootMargin: string;
  thresholds: readonly number[];

  elements: Set<Element>;

  constructor(
    public callback: IntersectionObserverCallback,
    options: IntersectionObserverInit = {}
  ) {
    this.root = options.root || null;
    this.rootMargin = options.rootMargin || "";
    this.thresholds =
      typeof options.threshold === "number"
        ? [options.threshold]
        : options.threshold ?? [];

    this.elements = new Set();
  }

  observe(target: Element): void {
    this.elements.add(target);

    this.callback(this.takeRecords(), this);
  }

  unobserve(target: Element): void {
    this.elements.delete(target);
  }

  takeRecords(): IntersectionObserverEntry[] {
    return [...this.elements].map<IntersectionObserverEntry>((target) => ({
      time: Date.now(),
      target,
      boundingClientRect: target.getBoundingClientRect(),
      intersectionRatio: 0,
      intersectionRect: target.getBoundingClientRect(),
      isIntersecting: false,
      rootBounds:
        this.root && "getBoundingClientRect" in this.root
          ? this.root.getBoundingClientRect()
          : null,
    }));
  }

  disconnect(): void {
    this.elements.clear();
  }
}
