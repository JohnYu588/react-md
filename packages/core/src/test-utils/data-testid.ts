/// <reference types="react" />

declare module "react" {
  interface HTMLAttributes<T>
    extends React.AriaAttributes,
      React.DOMAttributes<T> {
    "data-testid"?: string | number;
  }
}
