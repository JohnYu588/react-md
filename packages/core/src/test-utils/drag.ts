import { fireEvent } from "@testing-library/dom";
import type { MouseEvent } from "react";
import { wait } from "../utils/wait.js";

interface XYCoords {
  x: number;
  y: number;
}

interface BaseDragOptions {
  to?: XYCoords | Element;
  delta?: XYCoords;
  steps?: number;
  duration?: number;
}

type DragOptions = BaseDragOptions &
  ({ to: XYCoords | Element; delta?: never } | { delta: XYCoords; to?: never });

const getElementClientCenter = (element: Element): XYCoords => {
  const { left, top, width, height } = element.getBoundingClientRect();
  return {
    x: left + width / 2,
    y: top + height / 2,
  };
};

const getCoords = (elementOrCoords: Element | XYCoords): XYCoords =>
  "x" in elementOrCoords && "y" in elementOrCoords
    ? elementOrCoords
    : getElementClientCenter(elementOrCoords);

/**
 * This is internal for now. No idea if it's actually worthwhile though since
 * you have to mock the `getBoundingClientRect()` to work.
 *
 * @internal
 * @see https://github.com/testing-library/testing-library-docs/blob/c20f8a8931e9f0a158fda9f01bbea8cb45ed1651/docs/example-drag.mdx
 */
export async function drag(
  element: Element,
  options: DragOptions
): Promise<void> {
  const { delta, to: inTo, steps = 20, duration = 500 } = options;
  const from = getElementClientCenter(element);
  const to = delta
    ? {
        x: from.x + delta.x,
        y: from.y + delta.y,
      }
    : getCoords(inTo);

  const step: XYCoords = {
    x: (to.x - from.x) / steps,
    y: (to.y - from.y) / steps,
  };

  const currentEvent = {
    clientX: from.x,
    clientY: from.y,
    button: 0,
    altKey: false,
    ctrlKey: false,
    metaKey: false,
    shiftKey: false,
  } satisfies Partial<MouseEvent>;

  fireEvent.mouseEnter(element, currentEvent);
  fireEvent.mouseOver(element, currentEvent);
  fireEvent.mouseMove(element, currentEvent);
  fireEvent.mouseDown(element, currentEvent);
  fireEvent.dragStart(element, currentEvent);

  const delay = duration / steps;
  for (let i = 0; i < steps; i++) {
    currentEvent.clientX += step.x;
    currentEvent.clientY += step.y;
    await wait(delay);
    fireEvent.mouseMove(element, currentEvent);
  }

  fireEvent.mouseUp(element, currentEvent);
  fireEvent.dragEnd(element, currentEvent);
}
