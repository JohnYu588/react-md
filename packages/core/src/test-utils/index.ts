export * from "@testing-library/react";
export * from "@testing-library/user-event";
export * from "./mocks/match-media.js";
export * from "./render.js";
export * from "./utils/resize-observer.js";
