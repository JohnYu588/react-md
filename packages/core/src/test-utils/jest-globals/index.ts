export * from "./match-media.js";
export * from "./resize-observer.js";
export * from "./timers.js";
