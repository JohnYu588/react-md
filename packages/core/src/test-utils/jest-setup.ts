import { beforeEach } from "@jest/globals";
import { INTERACTION_CONFIG } from "../interaction/config.js";

beforeEach(() => {
  // set the mode to `none` in tests since ripples require
  // `getBoundingClientRect()` to create correct CSS. You'll either see warnings
  // in the console around invalid css values or `NaN`.
  INTERACTION_CONFIG.mode = "none";
});
