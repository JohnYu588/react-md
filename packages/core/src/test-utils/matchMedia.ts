import { jest } from "@jest/globals";
import { act } from "@testing-library/react";
import {
  DEFAULT_DESKTOP_LARGE_MIN_WIDTH,
  DEFAULT_DESKTOP_MIN_WIDTH,
  DEFAULT_PHONE_MAX_WIDTH,
  DEFAULT_TABLET_MIN_WIDTH,
} from "../media-queries/appSize.js";

const noop = (): void => {
  // do nothing
};

/**
 * @see {@link spyOnMatchMedia} instead
 * @internal
 * @remarks \@since 6.0.0
 */
export const BASE_MEDIA_QUERY_LIST: MediaQueryList = {
  media: "",
  matches: false,
  onchange: noop,
  addListener: noop,
  addEventListener: noop,
  removeEventListener: noop,
  removeListener: noop,
  dispatchEvent: () => false,
};

/** @remarks \@since 6.0.0 */
export type MatchMediaMatcher = (query: string) => boolean;

/**
 * @see {@link spyOnMatchMedia} for usage
 * @remarks \@since 6.0.0
 * @returns `true` for phone media queries
 */
export const matchPhone: MatchMediaMatcher = (query) =>
  query.includes(`${DEFAULT_PHONE_MAX_WIDTH}`);

/**
 * @see {@link spyOnMatchMedia} for usage
 * @remarks \@since 6.0.0
 * @returns `true` for tablet media queries
 */
export const matchTablet: MatchMediaMatcher = (query) =>
  query.includes(`${DEFAULT_TABLET_MIN_WIDTH}`);

/**
 * @see {@link spyOnMatchMedia} for usage
 * @remarks \@since 6.0.0
 * @returns `true` for desktop media queries
 */
export const matchDesktop: MatchMediaMatcher = (query) =>
  query.includes(`${DEFAULT_DESKTOP_MIN_WIDTH}`);

/**
 * @see {@link spyOnMatchMedia} for usage
 * @remarks \@since 6.0.0
 * @returns `true` for large desktop media queries
 */
export const matchLargeDesktop: MatchMediaMatcher = (query) =>
  query.includes(`${DEFAULT_DESKTOP_LARGE_MIN_WIDTH}`);

/**
 * @see {@link spyOnMatchMedia} for usage
 * @remarks \@since 6.0.0
 * @returns `true` for both desktop and large desktop media queries
 */
export const matchAnyDesktop: MatchMediaMatcher = (query) =>
  matchDesktop(query) || matchLargeDesktop(query);

/**
 * @remarks \@since 6.0.0
 */
export type MatchMediaSpiedFunction = jest.SpiedFunction<
  typeof window.matchMedia
> & {
  /**
   * @example
   * Default Behavior
   * ```tsx
   * const matchMedia = spyOnMatchMedia();
   * render(<Test />);
   *
   * // expect desktop results
   *
   * matchMedia.changeViewport(matchPhone);
   * // expect phone results
   * ```
   *
   * @example
   * Custom Act Behavior
   * ```tsx
   * const matchMedia = spyOnMatchMedia();
   * const { rerender } = render(<Test />);
   *
   * // expect desktop results
   *
   * matchMedia.changeViewport(matchPhone, false);
   * rerender(<Test key="new-key" />);
   *
   * // expect phone results
   * ```
   */
  changeViewport(matcher: MatchMediaMatcher, disableAct?: boolean): void;
};

/**
 * @example
 * Default Behavior
 * ```tsx
 * import { matchPhone, render, spyOnMatchMedia } from "@react-md/test-utils";
 *
 * const matchMedia = spyOnMatchMedia();
 * render(<Test />);
 *
 * // expect desktop results
 *
 * matchMedia.changeViewport(matchPhone);
 * // expect phone results
 * ```
 *
 * @example
 * Set Default Media
 * ```tsx
 * import { matchPhone, render, spyOnMatchMedia } from "@react-md/test-utils";
 *
 * const matchMedia = spyOnMatchMedia(matchPhone);
 * render(<Test />);
 *
 * // expect phone results
 * ```
 *
 * @remarks \@since 6.0.0
 */
export function spyOnMatchMedia(
  defaultMatch: MatchMediaMatcher = matchDesktop
): MatchMediaSpiedFunction {
  type Listener = (event: MediaQueryListEvent) => void;

  const listeners = new Map<string, Listener>();
  const matchMedia = jest
    .spyOn(window, "matchMedia")
    .mockImplementation((query) => ({
      ...BASE_MEDIA_QUERY_LIST,
      addEventListener(type: string, listener: Listener | EventListenerObject) {
        /* c8 ignore start */
        if (typeof listener !== "function" || type !== "change") {
          return;
        }
        /* c8 ignore end */

        listeners.set(query, listener);
      },
      removeEventListener(
        type: string,
        listener: Listener | EventListenerObject
      ) {
        /* c8 ignore start */
        if (typeof listener !== "function" || type !== "change") {
          return;
        }
        /* c8 ignore end */

        listeners.delete(query);
      },
      matches: defaultMatch(query),
    }));

  const changeViewport = (
    matcher: MatchMediaMatcher,
    disableAct = false
  ): void => {
    const update = (): void => {
      window.dispatchEvent(new Event("resize"));

      const event = new Event("change");
      listeners.forEach((listener, query) => {
        listener({
          ...event,
          media: "",
          matches: matcher(query),
        });
      });
    };
    if (disableAct) {
      update();
    } else {
      act(update);
    }
  };

  const mock = matchMedia as MatchMediaSpiedFunction;
  mock.changeViewport = changeViewport;

  return mock;
}
