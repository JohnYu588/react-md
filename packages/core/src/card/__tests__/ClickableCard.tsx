import { describe, expect, it, jest } from "@jest/globals";
import { createRef } from "react";
import { fireEvent, render, screen } from "test-utils";
import { ClickableCard } from "../ClickableCard.js";

describe("ClickableCard", () => {
  it("should apply the correct styling, HTMLAttributes, and allow a ref", () => {
    const ref = createRef<HTMLDivElement>();
    const props = {
      "data-testid": "card",
      ref,
      onClick: jest.fn(),
    } as const;
    const { rerender } = render(<ClickableCard {...props} />);

    const element = screen.getByTestId("card");
    expect(ref.current).toBeInstanceOf(HTMLElement);
    expect(ref.current).toBe(element);
    expect(element).toMatchSnapshot();

    rerender(
      <ClickableCard
        {...props}
        style={{ color: "white" }}
        className="custom-class-name"
      />
    );
    expect(element).toMatchSnapshot();
  });

  it("should not have a tabIndex while disabled", () => {
    render(<ClickableCard data-testid="card" disabled onClick={jest.fn()} />);
    const card = screen.getByTestId("card");
    expect(card).toHaveAttribute("aria-disabled");
    expect(card).not.toHaveAttribute("tabIndex");
  });

  it("should keyboard click correctly", () => {
    const onClick = jest.fn();
    render(<ClickableCard data-testid="card" onClick={onClick} />);

    const card = screen.getByTestId("card");
    expect(onClick).not.toHaveBeenCalled();

    fireEvent.keyDown(card, { key: " " });
    expect(onClick).toHaveBeenCalledTimes(1);

    fireEvent.keyDown(card, { key: "Enter" });
    expect(onClick).toHaveBeenCalledTimes(2);
  });

  it("should not fire the click event while disabled", () => {
    const onClick = jest.fn();
    render(<ClickableCard data-testid="card" onClick={onClick} disabled />);
    const card = screen.getByTestId("card");

    fireEvent.click(card);
    expect(onClick).toHaveBeenCalledTimes(0);

    fireEvent.keyDown(card, { key: " " });
    expect(onClick).toHaveBeenCalledTimes(0);

    fireEvent.keyDown(card, { key: "Enter" });
    expect(onClick).toHaveBeenCalledTimes(0);
  });
});
