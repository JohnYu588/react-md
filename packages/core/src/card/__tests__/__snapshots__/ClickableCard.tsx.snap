// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ClickableCard should apply the correct styling, HTMLAttributes, and allow a ref 1`] = `
<div
  class="rmd-box rmd-box--column rmd-box--align-stretch rmd-box--justify-stretch rmd-card rmd-card--shadowed rmd-card--clickable"
  data-testid="card"
  role="button"
  tabindex="0"
/>
`;

exports[`ClickableCard should apply the correct styling, HTMLAttributes, and allow a ref 2`] = `
<div
  class="rmd-box rmd-box--column rmd-box--align-stretch rmd-box--justify-stretch rmd-card rmd-card--shadowed rmd-card--clickable custom-class-name"
  data-testid="card"
  role="button"
  style="color: white;"
  tabindex="0"
/>
`;
